src/fundus/scraping/pipeline.py

```diff
@@ -62,6 +61,102 @@ def __init__(self, *scrapers: Scraper):
+            delay (Optional[Union[float, Delay]]): Set delay time between article batches. Defaults to None.
+            url_filter (Optional[URLFilter]): Set URLFilter. Defaults to None.
+            only_unique (bool): If true return only unique responses. Defaults to True.
+
+        Returns:
+            AsyncIterator[Article]: An iterator yielding objects of type Article.
+        """
+
+        def build_extraction_filter() -> Optional[ExtractionFilter]:
+            if isinstance(only_complete, bool):
+                return (
+                    None
+                    if only_complete is False
+                    else lambda extracted: not all(
+                        bool(v) if not isinstance(v, Exception) else False for k, v in extracted.items()
+                    )
+                )
+            else:
+                return only_complete
+
+        def build_delay() -> Optional[Delay]:
+            if isinstance(delay, float):
+
+                def constant_delay() -> float:
+                    return delay  # type: ignore[return-value]
+
+                return constant_delay
+            else:
+                return delay
+
+        def build_unique_url_filter() -> URLFilter:
+            return lambda url: url in response_cache
+
+        # build filters and delay. this is for readability and typeguard reasons
+        extraction_filter = build_extraction_filter()
+        unique_url_filter = build_unique_url_filter() if only_unique else None
+        final_delay = build_delay()
+
+        response_cache: Set[str] = set()
+
+        for scraper in self.scrapers:
+            for source in scraper.sources:
+                if url_filter:
+                    source.add_url_filter(url_filter=url_filter)
+                if unique_url_filter:
+                    source.add_url_filter(url_filter=unique_url_filter)
+
+        async_article_iterators: List[AsyncIterator[Optional[Article]]] = [
+            scraper.scrape(
+                error_handling=error_handling,
+                extraction_filter=extraction_filter,
+            )
+            for scraper in self.scrapers
+        ]
+
+        # we use this custom variant of interleave_longest in order to be able
+        # to delay the program flow between batches
+        async def _async_article_interleave_longest() -> AsyncIterator[Article]:
+            batches: AsyncIterator[Tuple[Optional[Article], ...]] = async_zip_longest(*async_article_iterators)
+            start_time = time.time()
+            async for batch in batches:
+                basic_logger.debug(f"Batch took {time.time() - start_time} seconds")
+                for value in batch:
+                    if value is not None:
+                        response_cache.add(value.html.responded_url)
+                        yield value
+                if final_delay:
+                    await asyncio.sleep(final_delay() - time.time() + start_time)
```

Comment:
According to the [asyncio.sleep](https://github.com/python/cpython/blob/aab6f7173a3b825599629dd6fa5cb7e477421595/Lib/asyncio/tasks.py#L631C37-L631C37), it's the same as when used with delay=0, but to avoid future confusion I will add a call to `max`.
That is a nice example of working by accident, while writing I didn't consider the very obvious case of execution time > delay time. I'm glad you found this.

Reply:
Interesting implementation detail. Thank's for the clarification :).